### PR TITLE
ci: use YAML references to define the build-image sha once

### PR DESCRIPTION
**Pull request by gordonsyme**
Removes duplication of the build-image SHA in `.circleci/config.yml`

Signed-off-by: Gordon Syme <gordonsyme@gmail.com>

### CODE REVIEW

**Comment by gordonsyme** [comment]
Ah, those CI jobs ran using my personal CircleCI plan which doesn't have configurable resources enabled. I'll remove my fork's project and force-push this commit to trigger a proper workflow.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,11 +1,15 @@
 # TODO(mattklein123): See if there is a way to put build SHA into a variable.
 # In the meantime, it's required to edit the SHA in this file as well as envoy_build_sha.sh.
 
+references:
```
Comment (mattklein123) [inline .circleci/config.yml]: Actually sorry can you delete the TODO comments above. Both of them are no longer needed!

Comment (gordonsyme) [reply .circleci/config.yml]: Sure thing.

**Comment by gordonsyme** [comment]
> Neat! TIL something new about YAML.

Yeah, it's pretty handy alright. If you ever get to the point of needing to set env-vars for the build image you can hoist the whole `image: ...` block into a reference like so

```yaml
references:
  envoy-build-image: &envoy-build-image
    image: envoyproxy/envoy-build:516cdcd55326648978f8db1fe3a774ec759f5a13
    environment:
      FOO: bar
      BAZ: qux

version: 2
jobs:
  release:
    docker:
      - *envoy-build-image
```

**Pull request merged**